### PR TITLE
Add cached `_split_line_selector` to avoid redundant parsing in `select_lines`

### PR DESCRIPTION
**Pull request by kenodegard**
### Description

A continuation of https://github.com/conda/conda-build/pull/5233 with inspiration from https://github.com/conda/conda-build/pull/5225.

When rendering a recipe we end up invoking `conda_build.metadata.select_lines` a number of times.

Within `select_lines` we first iterate over every line and split the line into the line content versus the selector and second we evaluate the selector to determine whether the line should be kept.

The first process of iterating over every line and splitting it should (and can) be a one time operation. This is gained by introducing a new `_split_line_selector` function which caches the parsed result for a given text.

The second process of evaluating the selectors can also be cached but needs to be cached differently. Instead of caching this globally (since selectors can change between rendering passes) we cache the selectors locally, so a given selector will only be evaluated once for a single pass of the text in question (it's common for a given selector to be used multiple times within a single file, e.g., [conda-forge's `conda_buid_config.yaml`](https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/main/recipe/conda_build_config.yaml)).

### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [x] Add / update necessary tests?
- [ ] ~Add / update outdated documentation?~

### CODE REVIEW

File: conda_build/metadata.py [hunk 1]
```diff
@@ -1664,7 +1690,6 @@ def build_id(self):
             raise RuntimeError(
                 f"Couldn't extract raw recipe text for {self.name()} output"
             )
-        raw_recipe_text = self.extract_package_and_build_text()
```
Comment (kenodegard) [inline conda_build/metadata.py:1667]: This is a duplicate invocation, look 5 lines above here

File: conda_build/metadata.py [hunk 2]
```diff
@@ -267,38 +267,64 @@ def eval_selector(selector_string, namespace, variants_in_place):
...
+                    if value:
+                        lines.append(line)
+                except Exception as e:
+                    sys.exit(
+                        f"Error: Invalid selector in meta.yaml line {i + 1}:\n"
+                        f"offending line:\n"
+                        f"{line}\n"
+                        f"exception:\n"
+                        f"{e.__class__.__name__}: {e}\n"
+                    )
```
Comment (beckermr) [inline conda_build/metadata.py:327]: These sys.exit calls in conda-build are very painful for downstream users of the code base as a library (i.e., conda-forge). The reason is that in the python exception hierarchy has the `SystemExit` exception raised by `sys.exit` and the standard `Exception` at the same level (https://docs.python.org/3/library/exceptions.html#exception-hierarchy). This means when one calls `sys.exit`, the invoking code cannot catch that exception without resorting to explicitly catching `SystemExit` as opposed to a normal `Exception`. The slight change in exception handling API breaks (at least my) expectations as a user of the code and has created quite a few mystery code crashes in the bot over the years. 

Can we use a standard exception here and then have the conda-build CLI code handle that exception and call `sys.exit`?

Comment (kenodegard) [reply conda_build/metadata.py:327]: I completely agree that we need to get away from these `exit` calls, just created https://github.com/conda/conda-build/issues/5242

**Comment by beckermr** [comment]
Have you tested this with the recipe @mbargull was working on? I'm wondering if the hotspots measured here are the same as those when big rerendering happens.

**Comment by beckermr** [comment]
Here is a timing test on rerendering

```zsh
(cf-dev) beckermr@finnegan ctng-compilers-feedstock % time conda-smithy rerender 
INFO:conda_smithy.configure_feedstock:Downloading conda-forge-pinning-2024.03.19.15.37.47
INFO:conda_smithy.configure_feedstock:Extracting conda-forge-pinning to /Users/beckermr/.cache/conda-smithy
INFO:conda_smithy.configure_feedstock:__pycache__ rendering is skipped
INFO:conda_smithy.configure_feedstock:README rendering is skipped
WARNING: Setting build platform. This is only useful when pretending to be on another platform, such as for rendering necessary dependencies on a non-native platform. I trust that you know what you're doing.
WARNING: Setting build arch. This is only useful when pretending to be on another arch, such as for rendering necessary dependencies on a non-native arch. I trust that you know what you're doing.
WARNING: No numpy version specified in conda_build_config.yaml.  Falling back to default numpy value of 1.23
Adding in variants from internal_defaults
Adding in variants from /Users/beckermr/.cache/conda-smithy/conda_build_config.yaml
Adding in variants from /Users/beckermr/Desktop/conda-forge/ctng-compilers-feedstock/recipe/conda_build_config.yaml
Adding in variants from argument_variants
INFO:conda_smithy.configure_feedstock:Re-rendered with conda-build 24.1.3.dev39, conda-smithy 3.32.0, and conda-forge-pinning 2024.03.19.15.37.47
INFO:conda_smithy.configure_feedstock:You can commit the changes with:

    git commit -m "MNT: Re-rendered with conda-build 24.1.3.dev39, conda-smithy 3.32.0, and conda-forge-pinning 2024.03.19.15.37.47"

INFO:conda_smithy.configure_feedstock:These changes need to be pushed to github!

conda-smithy rerender  475.86s user 79.66s system 88% cpu 10:27.78 total
(cf-dev) beckermr@finnegan ctng-compilers-feedstock % time conda-smithy rerender 
INFO:conda_smithy.configure_feedstock:__pycache__ rendering is skipped
INFO:conda_smithy.configure_feedstock:README rendering is skipped
WARNING: Setting build platform. This is only useful when pretending to be on another platform, such as for rendering necessary dependencies on a non-native platform. I trust that you know what you're doing.
WARNING: Setting build arch. This is only useful when pretending to be on another arch, such as for rendering necessary dependencies on a non-native arch. I trust that you know what you're doing.
WARNING: No numpy version specified in conda_build_config.yaml.  Falling back to default numpy value of 1.23
Adding in variants from internal_defaults
Adding in variants from /Users/beckermr/.cache/conda-smithy/conda_build_config.yaml
Adding in variants from /Users/beckermr/Desktop/conda-forge/ctng-compilers-feedstock/recipe/conda_build_config.yaml
Adding in variants from argument_variants
INFO:conda_smithy.configure_feedstock:Re-rendered with conda-build 24.1.2, conda-smithy 3.32.0, and conda-forge-pinning 2024.03.19.15.37.47
INFO:conda_smithy.configure_feedstock:You can commit the changes with:

    git commit -m "MNT: Re-rendered with conda-build 24.1.2, conda-smithy 3.32.0, and conda-forge-pinning 2024.03.19.15.37.47"

INFO:conda_smithy.configure_feedstock:These changes need to be pushed to github!

conda-smithy rerender  642.32s user 95.69s system 89% cpu 13:42.06 total
```

So the code is about 35% faster with this change on one of our slowest rerendering tasks. That is a lot less than the 10x in the performance tests.

**Comment by mbargull** [comment]
@kenodegard, thanks for taking a stab on this!
I haven't yet gotten to review this.
I could imagine this giving similar speed ups as my bolted-on approach for `select_lines` in gh-5225.
Having the line selection and per-variant-evaluation split (with the former in a global cache) could very well offset the ugly and costly "try to make `variant` hashable" dance I did in gh-5225.

----
@beckermr, thanks for trying this out already!
35% actually sounds good!
We have other parts that contribute much to the runtime overall.
For one, we have a big offset due to the unnecessary full-index creation (see gh-5154).
And then we do have the issue with conda-forge's large `conda_build_config.yaml` which is severely affected by the `.config.Config.copy`/`copy.deepcopy` issue mentioned in gh-5225.
Meaning, for `select_lines` alone, the timings look promising!

----
@kenodegard, I'll later try to strip the `arrow-cpp` example down to a test case we can use with your recent CodSpeed setup and update the tracking issue gh-5224 with the recent progress.

**Comment by kenodegard** [comment]
@beckermr @mbargull the new benchmarks results are in: https://github.com/conda/conda-build/pull/5237#issuecomment-1998627131

File: conda_build/metadata.py (hunk 2 above)
Comment (beeankha) [reply conda_build/metadata.py:328]: Are we keeping [this `sys.exit` call](https://github.com/conda/conda-build/pull/5237/files#diff-a3f5613bda9366e31149c65731327047d4e29204c9e0508fd55416115b39a6bdR322) for this PR but then changing it (along with the other 80+ instances) in a separate PR? Or should this particular instance of `sys.exit` be altered before we merge this work?

Comment (kenodegard) [reply conda_build/metadata.py:328]: I would like to leave this as is here and work on removing `sys.exit` separately

**Review by beeankha — APPROVED** [review]

File: conda_build/metadata.py (hunk 1 above)
Comment (kenodegard) [reply conda_build/metadata.py]: Split off into https://github.com/conda/conda-build/pull/5253

**Review by beeankha — DISMISSED** [review]

**Review by jezdez — APPROVED** [review]

**Pull request merged**